packages/eslint-plugin/src/rules/no-array-delete.ts

```diff
@@ -0,0 +1,90 @@
+import type { TSESLint, TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import { unionTypeParts } from 'tsutils';
+import * as ts from 'typescript';
+
+import * as util from '../util';
+
+export default util.createRule({
+  name: 'no-array-delete',
+  meta: {
+    hasSuggestions: true,
+    docs: {
+      description: 'Disallow delete operator for arrays',
+      recommended: 'strict',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      arrayDelete: 'Using the delete operator on an array is dangerous.',
+    },
+    schema: [],
+    type: 'problem',
+    fixable: 'code',
+  },
+  defaultOptions: [],
+  create(context) {
+    const parserServices = util.getParserServices(context);
+    const checker = parserServices.program.getTypeChecker();
+
+    return {
+      "UnaryExpression[operator='delete'] > MemberExpression[computed]"(
+        node: TSESTree.MemberExpressionComputedName & {
+          parent: TSESTree.UnaryExpression;
+        },
+      ): void {
+        const originalNode = parserServices.esTreeNodeToTSNodeMap.get(node);
+
+        const target = originalNode.getChildAt(0);
+        const key = originalNode.getChildAt(2);
+
+        const targetType = util.getConstrainedTypeAtLocation(checker, target);
+
+        if (!isTypeArrayTypeOrArrayInUnionOfTypes(targetType, checker)) {
+          return;
+        }
+
+        const keyType = util.getConstrainedTypeAtLocation(checker, key);
+
+        if (!isTypeNumberOrNumberLiteralOrNumberLikeType(keyType)) {
+          return;
+        }
+
+        context.report({
+          node,
+          messageId: 'arrayDelete',
+          suggest: [
```

Comment:
[Bug] This test case causes a suggested output that's a type error:

```ts
declare const arr: unknown[];
declare const i: number | string;

delete arr[i];
```

The suggestion fixer shouldn't suggest something that's a type error - so I think it shouldn't suggest anything at all in this case.